File: python/3_longest_substring.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_length = 0
        start = 0
        last_seen = {}

        for end, c in enumerate(s):
            if c in last_seen and last_seen[c] >= start:
                start = last_seen[c] + 1

            max_length = max(max_length, end - start + 1)
            last_seen[c] = end

        return max_length


if __name__ == "__main__":
    sol = Solution()

    print(sol.lengthOfLongestSubstring("abcabcbb"))

File: cpp/3_longest_substring.cpp
#include <string>
#include <unordered_map>

class Solution
{
	int lengthOfLongestSubstring(std::string s)
	{
		int							  maxLength = 0;
		int							  start = 0;
		std::unordered_map<char, int> index;

		for (int end = 0; end < s.length(); end++)
		{

			if (index.find(s[end]) != index.end() && index[s[end]] >= start)
				start = index[s[end]] + 1;
			maxLength = std::max(maxLength, end - start + 1);
			index[s[end]] = end;
		}
		return maxLength;
	}
};
